src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java

```diff
@@ -34,30 +34,33 @@
     PravegaTransactionWriterWorker(int sensorId, int eventsPerSec,
                                    int secondsToRun, boolean isRandomKey,
                                    int messageSize, Instant start,
-                                   ClientFactory factory, PerfStats stats,
-                                   String streamName,
-                                   long totalEvents,
+                                   PerfStats stats, String streamName,
+                                   long totalEvents, ClientFactory factory,
                                    int transactionsPerCommit) {
 
         super(sensorId, eventsPerSec, secondsToRun, isRandomKey,
-                messageSize, start, factory, stats, streamName, totalEvents);
+                messageSize, start, stats, streamName, totalEvents, factory);
 
         this.transactionsPerCommit = transactionsPerCommit;
         eventCount = new AtomicInteger(0);
         transaction = new AtomicReference<Transaction<String>>(producer.beginTxn());
     }
 
     @Override
-    public CompletableFuture writeData(String key, String data) throws TxnFailedException, IllegalStateException {
+    public CompletableFuture writeData(String key, String data) throws IllegalStateException {
         final Transaction<String> curTrans = transaction.get();
 
-        curTrans.writeEvent(key, data);
-        if (eventCount.incrementAndGet() >= transactionsPerCommit) {
-            eventCount.set(0);
-            curTrans.commit();
-            if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
-                throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
+        try {
+            curTrans.writeEvent(key, data);
+            if (eventCount.incrementAndGet() >= transactionsPerCommit) {
+                eventCount.set(0);
+                curTrans.commit();
+                if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
```

Comment:
This logic is not thread safe.

Reply:
eventCount is changed to transEventCount its Atomic interger and it increments rather than decrementing as per your last review comments. please let me know where exactly you are seeing an issue

Reply:
yes, I have moved  below two lines separate synchronized function: 
   
 private synchronized boolean transIncrementAndCompare() {
        if (transEventCount.incrementAndGet() >= transactionsPerCommit) {
            transEventCount.set(0);
            return true;
        }
        return false;
    }

Reply:
This doesn't fix the problem if transactionsPerCommit is 1.

Reply:
ok, then for time being for this PR, I have done some this:
        synchronized (this) {
                if (transEventCount.incrementAndGet() >= transactionsPerCommit) {
                    transEventCount.set(0);
                    curTrans.commit();
                    if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
                        throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
                    }
                }

synchronized count and commit.

while addressing https://github.com/pravega/pravega-benchmark/issues/16, I will revisit this logic to minimize the sync block